Create BaseService and incorporate into Ingress and Models

EdwinB12 opened the pull request:
Linked to #184 #181 #81 

Tasks: 

Compulsory things to do
- [x] Implement BaseService Class and get BaseIngress to Inherit
- [x] Create BaseModel class that inherits from BaseService
- [x] Move Arima over to BaseModel
- [x] Check WeatherIngress still works when inheriting from BaseService Class 

Extras
- [ ] Ensure user can override parameters, preferably from GUI 
- [ ] Add a run now functionality via frontend
- [ ] Design Scheduler system. This will likely be a PR in its own right as this needs integration with AzureFunctions.

EdwinB12 commented:
Remaining jobs on this PR

- [ ] Finish docstring for BaseModel.get_service_data() 
- [ ] Complete ArimaModel class so get_service_data() returns the list of endpoints/payload pairs. 
- [ ] Write tests for get_service_data() and __call__()

EdwinB12 commented:
The main things to check in this PR are:

- [ ] Base Class, in particular the call method and login. Could you check how the authentication is done. Think we might want to clean that up. (services/base.py)
- [ ] Implementation of Arima. Each individual function doesn't need as they're just copied over. However, please check the get_service_data() method and how the config is handled.  (models/arima.py)
- [ ] tests : Only need to check test_model_arima.py and test_services_base.py
- [ ] Examples and arima folder: I'm proposing removing these files. If you agree we don't need them then they don't need reviewing. 
- [ ] Nothing to do with WeatherIngress needs reviewing. The changes in those files are largely refactoring. 
- [ ] Please give the ReadMe a quick look over 

Thanks!

mhauru commented on dtbase/models/README.md:
````diff
@@ -0,0 +1,45 @@
+```
+For models, the endpoints that likely need to be returned are:
+
+- /model/insert-model
````
Maybe wrap the endpoint paths in `backticks`

mhauru commented on dtbase/models/README.md:
```diff
@@ -0,0 +1,45 @@
+For models, the endpoints that likely need to be returned are:
+
+- /model/insert-model
+- insert-model-scenario
```
This in the path under it are missing the `/model/` part

mhauru commented on dtbase/models/arima.py:
```diff
@@ -0,0 +1,550 @@
+
+# set up logging
+logger = logging.getLogger(__name__)
+logging.basicConfig(stream=sys.stdout, level=logging.INFO)
```
Does this line change the logging config globally whenever someone imports this module? I always forget how this logging config stuff works. If yes, that could perhaps be changed. In general, I think logging config setting is best left for scripts rather than library-like modules.

EdwinB12 replied:
Agreed

EdwinB12 replied:
I've actually removed all this formatting of the logs here. I think its probabaly uneccessary

mhauru commented on dtbase/models/arima.py:
```diff
@@ -0,0 +1,550 @@
+        Report errors if the input data is not a pandas Series indexed by timestamp.
+        """
+        if not isinstance(data.index, pd.DatetimeIndex):
+            logger.error(
```
Rather than logging an error and raising a `ValueError`, could we just raise `raise ValueError("blahblah explanation")`?

EdwinB12 replied:
Yup, cleaner

mhauru commented on dtbase/models/arima.py:
```diff
@@ -0,0 +1,550 @@
+            )
+            raise ValueError
+
+    def get_config_from_config_files(self) -> ConfigArima:
```
Could this be renamed to something like `set_config_defaults`? It only reads values from the files for configuration variables that are not yet set in `self.config`.

mhauru replied:
I don't think this function returns anything, hence the type annotation should probably be `None`.

mhauru commented on dtbase/models/arima.py:
```diff
@@ -0,0 +1,550 @@
+        """
+        arima_config = self.config["arima"]
+        if arima_config.hours_forecast <= 0:
+            logger.error(
```
Same as above, could we just raise an exception with a string argument?

mhauru commented on dtbase/models/examples/run_arima.py:
I'm not sure how to best do the directory structure, but I think it could be nice to have some way to run the model locally. Or would the way to do it be to spin up the Azure function app thing locally and call that? I guess that would work too.

mhauru replied:
Yeah that could work. What about parameter passing? As a user, I'm thinking that it's nice to be able to do a test run locally, but I'm not sure how I would tell the model what to run on.

EdwinB12 replied:
Lets address this in #212

mhauru commented on dtbase/models/examples/run_hodmd.py:
If we are not overhauling the HODMD model more generally yet, I wonder if this file should be kept somewhere? Unsure about this.

EdwinB12 replied:
Yer think you might have a point here. Guess this depends on whether we want to keep hodmd as a method in dtbase?

EdwinB12 replied:
I'm pretty sure we do, so maybe best to return this file to version control. . however you do that

mhauru commented on dtbase/models/examples/test_hodmd.ipynb:
What are your thoughts on these sorts of notebooks? I generally dislike notebooks, and having them mixed in with the library feels wrong. On the other hand, I wonder if we will miss some of the model documentation in this file when we get to actually understanding HODMD.

EdwinB12 replied:
I think they have no place in the module. Instead, I think we should have a directory caller "examples" at repository level that has notebooks that demonstrate how to do basic tasks like deploy an instance or write a custom ingress, etc.

mhauru commented on dtbase/services/base.py:
```diff
@@ -0,0 +1,218 @@
+class BaseService:
+    """
+    Base class for all services. Users inherit this class to write custom modelling and
+    data ingress services. The user should implement the following methods:
```
The docstring seems unfinished.

mhauru commented on dtbase/services/base.py:
```diff
@@ -0,0 +1,218 @@
+        """
+        self.access_token = login(username, password)[0]
+
+    def get_service_data(self) -> None:
```
I wonder if the return type annotation here should be what the method should return once it's been implemented. So something like `list[tuple[str, dict]]`. That would help document how this method is expected to be used.

mhauru commented on dtbase/services/base.py:
```diff
@@ -0,0 +1,218 @@
+        dt_user_password: Optional[str] = None,
+    ) -> List[Response]:
+        """
+        Upload data to the database via the backend via a post request.
```
A slight repharsing of the double-via might make this clearer.

mhauru commented on dtbase/services/base.py:
```diff
@@ -0,0 +1,218 @@
+        data_pairs = self.get_service_data(**kwargs)
+        return self.post_service_data(data_pairs, dt_user_email, dt_user_password)
+
+    def schedule(self) -> None:
```
I think this logic may eventually end up living somewhere else. I think `BaseService` is essentially what someone implementing a service will subclass, to interface with DTBase. I think the scheduling logic is likely to instead live in the backend. Or to put this another way `BaseService` would deal with "what happens when someone asks this service to run" and the backend would deal with "when should each service be run".

EdwinB12 replied:
Agreed, I'll remove it for now as it isn't implemented

mhauru commented on dtbase/services/base.py:
```diff
@@ -0,0 +1,218 @@
+class BaseIngress(BaseService):
+    """
+    Class to be inherited for custom data Ingress.
+    The user should implement the call method.
```
This seems out of date, I don't think call is the one to implement.

mhauru commented on dtbase/services/base.py:
```diff
@@ -0,0 +1,218 @@
+            ("/model/insert-model-predictions", MODEL_PREDICTIONS_PAYLOAD),
+        ]
+
+        Below is an example return value for inserting a model and its parameters.
```
This seems unfinished.

mhauru commented on dtbase/tests/generate_synthetic_data.py:
```diff
@@ -230,3 +234,13 @@ def generate_weather_forecast() -> pd.DataFrame:
     return df
+
+
+if __name__ == "__main__":
```
Was this just for debugging?

mhauru commented on dtbase/tests/test_model_arima.py:
```diff
@@ -41,12 +47,13 @@ def test_arima_get_temperature(
 @pytest.mark.skipif(not DOCKER_RUNNING, reason="requires docker")
 def test_arima_get_humidity(conn_backend: None, session: Session) -> None:
-    insert_trh_readings(session)
+    insert_trh_readings(session, add_noise=False)
```
Is the `add_noise=False` still needed now that the seed is fixed? Same question for several calls below.

EdwinB12 replied:
Correct. I'm still going to keep it as an option in the function because it could be useful but I won't call it here.

EdwinB12 replied:
By correct, I mean add_noise isn't need I don't think

mhauru commented on dtbase/tests/test_model_arima.py:
```diff
@@ -116,20 +128,79 @@ def test_arima_prepare(conn_backend: None, session: Session) -> None:
     values = prepared_data["TRH1"]["Temperature"]
-    mean_forecast, conf_int, metrics = arima_pipeline(values, config["arima"])
+
+    print(values)
```
A debugging print?

mhauru reviewed (changes requested):
Good stuff! Left some minor comments and requests scattered around the files.

EdwinB12 commented on dtbase/README.md:
```diff
@@ -21,10 +21,18 @@ The endpoints are grouped into `location`, `model`, and `sensor`, and within eac
+## Ingress
 
-This is where code to retrieve data from e.g. external APIs will live.   The scripts here will retrieve such data and use the backend API endpoints to insert it into the database.
+This is where code for specific data ingress is located. Data ingress is the act of pulling in data from another source such as an external API or database and inserting into the dtbase database via the backend. Instructions for how to write custom models can be found [here](ingress/README.md).
```
Nice catch

EdwinB12 commented on dtbase/functions/arima/__init__.py:
```diff
@@ -4,7 +4,7 @@
 from azure.functions import HttpRequest, HttpResponse
 
-from dtbase.models.arima.run_pipeline import run_pipeline
+from dtbase.models.arima_old.run_pipeline import run_pipeline
```
Ahh yes it is. I haven't tested much to do with the azure functions because we're probably going to redesign it. So for now, I will replace this with that it should now be but I won't test it extensively for now

EdwinB12 commented on dtbase/models/README.md:
```diff
@@ -0,0 +1,45 @@
+## BaseModel
+
+The `BaseModel` class has all the general purpose tools for interacting with the backend (and TODO: Azure Functions). It inherits from `BaseService`.
+For a custom models, the user should create their own custom DataModelClass inheriting from the `BaseModel` class.
```
That is a find and replace error I think. Good spot and agree.

EdwinB12 commented:
Thanks for the review @mhauru , I've addressed your comments. 

Now, the tests are failing. I wonder if its not worth testing for exact Arima results in the tests? The tests didn't before this and its just causing tests to fall over on 3rd decimal places. .

mhauru commented:
Thanks Edwin, I'll try to take a look later today.

> Now, the tests are failing. I wonder if its not worth testing for exact Arima results in the tests? The tests didn't before this and its just causing tests to fall over on 3rd decimal places.

Yeah, if it's turning into a pain then we can give this up. Maybe we can do some other sanity checks of the results instead (or do other tests do that already?). You could also reaaally reduce the precision on the checks, but I don't know if we have a guarantee that that won't ever fail, and indeterministically failing tests are nasty.

mhauru commented on dtbase/services/base.py:
```diff
@@ -14,8 +14,9 @@
-    data ingress services. The user should implement the following methods:
+    Base class for all services. This class should provide all generic methods for any
+    services such as ingress and models. BaseModel and BaseIngress inherit from this
+    method.
```
```suggestion
    class.
```

EdwinB12 commented:
Changes made, just need an approval @mhauru

EdwinB12 commented:
closes #184

mhauru reviewed (approved):
Thanks Edwin, great work!
